File: src/agent.ts
import { spawn, ChildProcess } from "child_process";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { GoogleGenAI, FunctionDeclaration } from "@google/genai";
import * as path from "path";
import { fileURLToPath } from "url";

// Calculate __dirname
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Initialize Google GenAI
const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
if (!GEMINI_API_KEY) {
  console.error("Missing GEMINI_API_KEY environment variable. AI won't work!");
}
const ai = new GoogleGenAI({ apiKey: GEMINI_API_KEY || "dummy" });

export class ADKAgent {
  private client: Client;
  private tools: FunctionDeclaration[] = [];
  private mcpToolsList: any[] = [];

  constructor() {
    this.client = new Client(
      {
        name: "adk-weather-agent",
        version: "1.0.0",
      },
      {
        capabilities: {},
      }
    );
  }

  async initialize() {
    console.log("Connecting Agent to MCP Server...");
    
    // We determine whether we are running as compiled JS (.js) or TypeScript (.ts)
    // This allows it to work from dev environment (tsx) and production Docker (node dist/).
    const isCompiled = __filename.endsWith('.js');
    const serverFileName = isCompiled ? "mcp-server.js" : "mcp-server.ts";
    const command = isCompiled ? "node" : "npx";
    const args = isCompiled ? [path.resolve(__dirname, `./${serverFileName}`)] : ["tsx", path.resolve(__dirname, `./${serverFileName}`)];
    
    const transport = new StdioClientTransport({ command, args });
    
    await this.client.connect(transport);
    
    console.log("MCP Connected. Discovering tools...");
    const response = await this.client.listTools();
    this.mcpToolsList = response.tools;
    console.log(`Discovered ${this.mcpToolsList.length} tools.`);

    // Map MCP Tools to Gemini FunctionDeclarations
    this.tools = this.mcpToolsList.map((tool) => {
      return {
        name: tool.name,
        description: tool.description,
        parameters: {
            ...tool.inputSchema
        }
      } as unknown as FunctionDeclaration;
    });
  }

  async query(userPrompt: string): Promise<string> {
    console.log(`Sending query to Gemini: "${userPrompt}"`);

    // Initiate multi-turn chat to handle function calls
    // Note: ensure model is gemini-2.5-flash with sufficient context capabilities
    let chat;
    try {
        chat = ai.chats.create({
          model: "gemini-2.5-flash",
          config: {
            tools: [{ functionDeclarations: this.tools }],
          },
        });
    } catch (e: any) {
         return `Failed to initialize Gemini. Did you export GEMINI_API_KEY? Error: ${e.message}`;
    }

    try {
      let response = await chat.sendMessage({ message: userPrompt });

      // Check if Gemini requested a function call
      if (response.functionCalls && response.functionCalls.length > 0) {
        for (const call of response.functionCalls) {
          console.log(`Gemini called Tool: ${call.name}`);
          
          // Execute tool directly on the unified MCP Client Transport connecting to our background Server Process
          const result = await this.client.callTool({
            name: call.name || "",
            arguments: call.args || {},
          });

          const content = result.content as { type: string; text: string }[];
          const toolData = content[0]?.text || "No data returned";
          console.log(`MCP Tool execution successful. Result length: ${toolData.length}`);

          // Send function result back to Gemini so it can generate final answer
          response = await chat.sendMessage({
            message: [{
                functionResponse: {
                    name: call.name || "",
                    response: { result: JSON.stringify(toolData) }
                }
            }]
          });
        }
      }
      
      return response.text || "No response generated";
    } catch (e: any) {
      console.error(e);
      throw e;
    }
  }

  async shutdown() {
      await this.client.close();
  }
}

File: src/agent.py
import sys
import json
import asyncio
from google.adk.agents import Agent

# Tool connection to the Node.js MCP server
async def get_weather(latitude: float, longitude: float) -> str:
    """
    Get the weather data from Google Cloud Weather API for a specific location.
    Args:
        latitude: Latitude coordinate
        longitude: Longitude coordinate
    """
    try:
        from mcp.client.stdio import stdio_client, StdioServerParameters
        from mcp.client.session import ClientSession
        
        # We spawn the existing Node.js MCP Server which contains Google API logic
        server_params = StdioServerParameters(
            command="node",
            args=["dist/mcp-server.js"]
        )

        async with stdio_client(server_params) as (read, write):
            async with ClientSession(read, write) as client:
                await client.initialize()
                
                result = await client.call_tool(
                    name="get_weather",
                    arguments={"latitude": latitude, "longitude": longitude}
                )
                
                # Assume standard MCP tool content response
                return str(result.content[0].text)
    except Exception as e:
        return f"Error executing MCP tool: {str(e)}"

async def main():
    if len(sys.argv) < 2:
        print(json.dumps({"error": "No prompt provided"}))
        sys.exit(1)

    prompt = sys.argv[1]

    try:
        # Initializes Google ADK Agent
        # Uses GEMINI_API_KEY from environment variables seamlessly
        agent = Agent(
            name="cloud_weather_copilot",
            model="gemini-2.5-flash",
            tools=[get_weather]
        )
        
        # Run ADK Orchestration
        response_chunks = []
        async for chunk in agent.run_async(prompt):
            # Check if chunk has a content attribute (like LiteLLM ModelResponse) or is string
            if hasattr(chunk, 'content'):
                response_chunks.append(str(chunk.content))
            else:
                response_chunks.append(str(chunk))
        
        response = "".join(response_chunks)

        # Output strictly structured JSON for the Express proxy index.ts to capture
        # Google ADK agent.run may return a string or object. str() cast is safest.
        print(json.dumps({"response": str(response)}))

    except Exception as e:
        print(json.dumps({"error": str(e)}))

if __name__ == "__main__":
    asyncio.run(main())
